### ISSUE TITLE
Performance issue migrating from TBB to oneTBB

### ISSUE THREAD

**Issue by JCLeprevost** [oneapi-src/oneTBB#1061]
Hi,
I started migrating our core algorithms from TBB to oneTBB and I am struggling to achieve the same performance as before. All our algorithms used the task interface and made intensive use of continuation and task recycling. Here is one of the simple case of our algorithms using TBB and the current replacement using oneTBB. The oneTBB version runs between 10-20% slower.

![image](https://user-images.githubusercontent.com/86110507/230890881-23dd7ecd-48a0-4323-8e97-11b1fb327998.png)

I am wondering if the performance lost is not due to a larger amount of tasks created using the oneTBB version. And if so, is there a way to use oneTBB to replicate the way the TBB code was operating.

Thanks, JC

**Comment by JCLeprevost** [comment]
I have tried using task groups and even though it is slightly better, the performance at the end is still far away from what I could achieve with the recycle as continuation that I have shown above.
One of the big issues I face is that I cannot replicate the "recycle_as_child( root )" in oneTBB because the operator() has a const and it does not allow changing local task parameters before recycling. While the task* execute() allowed me to do so.
My algorithms rely on a ternary tree with 2 branches that can run concurrently and a 3rd branch that has dependencies with the 2 others and thus need to be processed before as a root of the 2 others. I can't simply throw all the tasks without any temporal processing order because it would produce random results at worse if the "next" branch is executed concurrently with the others, or produce non deterministic results which is not an option for the kind of technology I develop.

We are quite stunned by the radical change of TBB and the deprecation of the task interface. To give you an idea of the impact it has on us, I ran a simulation for a 45 seconds movie shot that we kept as a benchmark because of the complexity of the multilayer clothing and the amount of collision stressing multiple layers of nested parallelism and the processing of millions of constraints both authored or dynamically generated. We handle cases that are 100 times bigger, but this case provide a benchmark in less than 1 min and not 10-15 min.
The engine using Tbb would run the shot in 21 seconds on a good workstation. The oneTBB version takes 31 seconds. Our collision detection performance  is 50% slower with oneTBB, our solver performance is 25% slower with oneTBB.
This is a huge problem for us. Because not only the deprecation is costing a massive amount of engineering time rewriting every single core algorithms to adapt to the radical change, but also we end up with a technology that is noticeably underperforming against its previous iteration. And performance is what makes our engine attractive.

I have gone through the migration tutorial and examples. And I did not find it as comprehensible as the Tbb documentation and examples used to be. It would be so much better to have very simple examples focussed on the oneTBB API itself and not have it drowned in a complicated problem and wrapped with unreadable lambda expressions with minimalistic comments. 
The simple algorithm I posted re-rooting as continuation and recycling as child of continuation was actually a TBB example showing a progressive improvement of a divide and conquer pattern if I remember properly. And yet, I could not find any oneTBB examples tackling such a versatile and flexible design and showing how to progressively improve its performance.

Any suggestions of how I could replicate the original task example I posted with one TBB would be really appreciated this points. Otherwise, we might have to look for alternatives to oneTBB as we can't downgrade our libraries in terms of execution time.

Thanks, JC

**Comment by pavelkumbrasev** [comment]
It should be possible to rewrite this example based on the Migration Guide. However, I'm not so familiar with the TBB Task API, but I'd be happy to try and rewrite this example (I have several ideas already).
May I ask you for an executable reproducer? This can make experiments easier.

**Comment by JCLeprevost** [comment]
I can't provide any library code or tests we run because assets are under NDA. The screen shot was from an existing code stripped out of most parameters and function calls too, just to expose the skeleton of the algorithm. 

So I went to find the tbb example that did show the allocate continuation and recycle as child of continuation.

This TBB library has examples of recycle child as continuation. In both test all fibonacci and task tree_sum. 
https://github.com/oneapi-src/oneTBB/releases/tag/v2020.3

I did not manage to find an example that uses a continuation task and recycle as child of continuation like I shown on my screen shot. But being able to replicate the performance of the Optimized parallel sum tree with oneTBB API following the same logic would be a big win.

**Comment by pavelkumbrasev** [comment]
I have rewrote the Fibonacci example with oneTBB I hope it would help. I tried to do it in several steps (like in the document you sent) 
**Basic version:**
```
#include "tbb/task_group.h"

long serial_fib(int n) {
    return n < 2 ? n : serial_fib(n - 1) + serial_fib(n - 2);
}

long simple_tg_parallel_fib(int n) {
    if (n < /*cutoff = */ 16) {
        return serial_fib(n);
    } else {
        long x, y;
        tbb::task_group tg;
        tg.run([&] { x = simple_tg_parallel_fib(n - 1); });
        tg.run([&] { y = simple_tg_parallel_fib(n - 2); });
        tg.wait();
        return x + y;
    }
}
```
**Version with continuation:**
```
#include "tbb/task_group.h"
#include <atomic>

long serial_fib(int n) {
    return n < 2 ? n : serial_fib(n - 1) + serial_fib(n - 2);
}
struct fib_continuation {
    fib_continuation(long& sum_, fib_continuation* parent_, tbb::task_group& tg_) : sum(sum_), parent(parent_), tg(tg_) {}

    fib_continuation(const fib_continuation&) = delete;

    void operator() () const {
        sum = x + y;
        if (parent && --parent->done == 0) {
            tg.run(std::ref(*parent));
        }
    }

    long& sum;
    fib_continuation* parent;
    tbb::task_group& tg;

    long x{};
    long y{};
    std::atomic<int> done{2};
};

struct fib_computation {
    fib_computation(long* sum_, int n_, tbb::task_group& tg_, fib_continuation* c)
        : sum(sum_), n(n_), tg(tg_), continuation(c)
    {}

    void operator() () const {
        if (n < /*cutoff = */ 16) {
            *sum = serial_fib(n);
            if (continuation && --continuation->done == 0) {
                tg.run(std::ref(*continuation));
            }
        } else {
            fib_continuation* cont = new fib_continuation(*sum, continuation, tg);
            tg.run(fib_computation{&cont->x, n - 1, tg, cont});
            tg.run(fib_computation{&cont->y, n - 2, tg, cont});
        }
    }

    long* sum;
    int n{};
    tbb::task_group& tg;
    fib_continuation* continuation;
};

long cont_tg_parallel_fib(int n) {
    long sum{};
    tbb::task_group tg;
    tg.run_and_wait(fib_computation{&sum, n, tg, nullptr});

    return sum;
}
```
**Version with continuation + recycle:**
```
#define __TBB_PREVIEW_TASK_GROUP_EXTENSIONS 1
#include "tbb/task_group.h"

#include <atomic>

long serial_fib(int n) {
    return n < 2 ? n : serial_fib(n - 1) + serial_fib(n - 2);
}

struct fib_continuation {
    fib_continuation(long& sum_, fib_continuation* parent_, tbb::task_group& tg_) : sum(sum_), parent(parent_), tg(tg_) {}

    fib_continuation(const fib_continuation&) = delete;

    void operator() () const {
        sum = x + y;
        if (parent && --parent->done == 0) {
            tg.run(std::ref(*parent));
        }
    }

    long& sum;
    fib_continuation* parent;
    tbb::task_group& tg;

    long x{};
    long y{};
    std::atomic<int> done{2};
};

struct fib_computation_with_recycle {
    fib_computation_with_recycle(long* sum_, int n_, tbb::task_group& tg_, fib_continuation* c)
        : sum(sum_), n(n_), tg(tg_), continuation(c)
    {}

    tbb::task_handle operator() () const {
        if (n < /*cutoff = */ 16) {
            *sum = serial_fib(n);
            // last reference was removed
            if (continuation && --continuation->done == 0) {
                tg.run(std::ref(*continuation));
            }
            return tbb::task_handle{};
        } else {
            fib_continuation* cont = new fib_continuation(*sum, continuation, tg);
            tg.run(fib_computation_with_recycle{&cont->x, n - 1, tg, cont});

            // Recycle task
            sum = &cont->y;
            n = n - 2;
            continuation = cont;
            return tg.defer(*this);
        }
    }

    mutable long* sum;
    mutable int n{};
    tbb::task_group& tg;
    mutable fib_continuation* continuation;
};

long rec_cont_tg_parallel_fib(int n) {
    long sum{};
    tbb::task_group tg;
    tg.run_and_wait(fib_computation_with_recycle{&sum, n, tg, nullptr});

    return sum;
}
```

**Comment by JCLeprevost** [comment]
Thank you so much. I will give it a try asap.
In the meanwhile I have tried to use flow graphs but this was not successful because the dependencies are between sub hierarchies in my problem and not just between nodes with single messaging edges. 
The recycle with continuation seems to be by far the best algorithm for the type of problem I am dealing with. 

Once again, thank you very much for your help. I will keep you updated with performance comparisons as soon as I have adapted this to our simulation engine.

**Comment by pavelkumbrasev** [comment]
Thanks, I'll be waiting for the results!

**Comment by JCLeprevost** [comment]
Hi Pavel,

I have been focussing on the continuation version you posted. I decided to put the recycling version on hold for now because it makes use of preview features and also make intensive use of mutable.

I could adapt your continuation algorithm in our collision detection which has a comparable reduce strategy but with up to 4 children instead of 2. I assume the done atomic replicates tbb reference count. But I have miserably failed to do so with our core solvers which relate to the initial algorithm I posted.
![image](https://user-images.githubusercontent.com/86110507/231722645-96e4da76-7226-49a7-b8bc-e398fc6b9d58.png)
In this algorithm Jobs describe a ternary tree. Each branching tasks may have a Next child, a Right and a Left concurrent children. Right and Left can execute in parallel but Next has to wait as a continuation for Left and Right to be finished. So when I tried to make this work with your continuation algorithm, I did not manage to make the continuation and the computation be the same structure like in the TBB algorithm with the task interface.
I did get something running with a computation and continuation structures by making the continuation spawn a computation. But the behavior is not deterministic which means I still failed since next is executing concurrently with right or left.
The performance was comparable to the old TBB version, which is the encouraging result so far.

I must admit I am quite demoralised by how complexe it is in one BB to achieve something that was simple to comprehend with the task interface in old TBB. We are used to adapt to interface deprecations, but it is quite frustrating to have to adapt to a functionality deprecation, because as a small business, we end up spending a lot of time rewriting something that worked well before without getting any additional benefits in terms of performance or result quality. This migration brings no value to the technology other than preventing the consequences of the task interface deprecation.

Anyway, I am not giving up yet. There has to be a way to make the continuation and the computation be the same structure.

One again, I really appreciated the effort you put in helping with this.
Cheers, JC

**Comment by JCLeprevost** [comment]
Thanks.

The ternary clustering is a form of divide and conquer. The idea is that we have a pool of constraints that would affect a group of objects and we split the pool iteratively in 3 groups. It is like drawing a demarcation line with objects on the left and objects on the right. Then we move all the constraints that affect objects that are on both sides of the line in a Branch called Next, the constraints affecting  the right objects are in the right pool and the ones affecting the left objects in the left pool. So at every iteration we have the next branch that has to wait for everything else to be recursively done, and we also split the work in 2 tasks that are concurrent each time we traverse the tree up to the leaves. The granularity is handled by the leaves containing N constraints.

This algorithm has 2 advantages:
-One of them is that the clustering can be created incredibly efficiently while processing the constraints. This is extremely precious when applied to an iterative problem as you build the tree once and reuse it multiple times.
-The second advantage is related to memory bandwidth. Due to the nature of the recursive partitioning the content of the leaves are sharing the same objects and thus prefetch data for one another. Compared to a very good graph colouring algorithm applied to the same problem, this algorithm demolishes the colouring by being 20-35% faster on CPU that is.

I tried to use one TBB graph but it was not fit for the problem since the temporal dependencies would need lots of edges. But I discovered this part of tbb that I never used before and I must admit, this is really good API, easy, comprehensible,... I plan to make use of it in the future for other things than collision detection and solve.

JC

**Comment by pavelkumbrasev** [comment]
There is an implementation of code example that you showed on oneTBB. (Unfortunately, I can not test it only compile)
```cpp
struct Map {};

struct Triple {
    int Next() const { return 0; }
    int Right() const { return 0; }
    int Left() const { return 0; }
};

struct Bucket { static constexpr int Mask = 1; };

struct base_continuation {
    base_continuation(base_continuation* parent_, int child_counter_, tbb::task_group& tg_) : m_parent(parent_), m_child_counter(child_counter_),  m_tg(tg_)
    {}

    base_continuation(const base_continuation&) = delete;

    void operator() () const {
        if (m_parent && --m_parent->m_child_counter == 0) {
            m_tg.run(std::ref(*m_parent));
        }
        // delete this;  // free memory
    }

    base_continuation* m_parent;
    std::atomic<int> m_child_counter;
    tbb::task_group& m_tg;
};


template <typename T>
void parallel_functor(const T& obj);

struct parallel_continuation : base_continuation {
    parallel_continuation(Map& map, std::int32_t fork, base_continuation* parent_, int child_counter_, tbb::task_group& tg_) : base_continuation(parent_, child_counter_, tg_), m_map(map), m_fork(fork)
    {}

    void operator() () const {
        parallel_functor(*this);
        // delete this;  // free memory
    }
    
    Map& m_map;
    std::int32_t m_fork;
};

struct parallel_computation {
    parallel_computation(Map& map, std::int32_t fork, base_continuation* parent_, tbb::task_group& tg_) : m_map(map), m_fork(fork), m_parent(parent_), m_tg(tg_) {}

    void operator() () const {
        parallel_functor(*this);
    }

    Map& m_map;
    std::int32_t m_fork;

    base_continuation* m_parent;
    tbb::task_group& m_tg;
};

template <typename T>
tbb::task_handle parallel_functor(const T& obj) {
    const Triple& job = obj.m_map.Jobs()[obj.m_fork];

    if (job.Next()) {
        base_continuation* continuation = job.Next() != Bucket::Mask ? new parallel_continuation(obj.m_map, obj.m_fork + job.Next(), obj.parent, 2, obj.m_tg) : new base_continuation(obj.parent, 2, obj.m_tg);
        obj.m_tg(parallel_computation{obj.m_map, obj.m_fork + job.Right(), continuation, obj.m_tg});
        obj.m_tg(parallel_computation{obj.m_map, obj.m_fork + job.Left(), continuation, obj.m_tg});
    } else {
        // Actually done computation
        if (obj.m_parent && --obj.m_parent->m_child_counter == 0) {
            obj.m_tg.run(std::ref(*obj.m_parent));
        }
    }
}
``` 
Let me know if I missed something important.
Regarding the Flow Graph in would be interesting to know performance comparing to pure task API it can show in your use case.

**Comment by JCLeprevost** [comment]
I ll give it a try this evening and feedback as soon as I get a result. 

The flow graph test I did was slow because I built the graph nodes and edges at every iterations before running it. Also it was not suitable for this specific solve algorithm since the children of right and Left would run concurrently with Next because it is not a time dependency scheduling but a message propagation that decide when tasks get executed.  But I did clock the build and execution: the graph build was taking 65% of the time and the execution 35% on average. So for a problem where the graph could be built at initialisation time and executed at run time iteratively, the flow graph did look to me like an elegant and very competitive solution. I would say that its execution performance alone was comparable to the pure task API in contrast with the task group that worked like the simple fib algorithm you suggested. That task group algorithm actually works because it would execute Next, then use a local task group for Right and Left. But it is 30-50% slower than the pure task API algorithm I posted.

**Comment by JCLeprevost** [comment]
Hi Pavel,

I managed to adapt this algorithm and it does the right thing.
I had to remove the parallel_functor that had no body, replace tbb::task_handle by void in the second functor and make it a static function. I also needed to use a virtual operator() () const from base and parallel continuation so that the correct operator get executed. I don't think there was any other issues.

The code produces the right result and is deterministic.

Now, I tested the performance with both our character tutorial which is a catwalk with 2 low resolution garments and our benchmark shot with 18 layers of cloth.

On the low small problem:
-The simple task group approach takes 0.05 seconds solve per frame.
-The last continuation algorithm takes 0.05 seconds solve per frame.
-The previous TBB task algorithm takes 0.04 seconds solve per frame.
So on small problems the old TBB is still 20% faster than one TBB algorithms we tried here.

On the benchmark scenario:
- Simple task group approach takes 0.057s per solve update.
- The continuation algorithm takes 0.043s per solve update.
- The previous TBB task algorithm takes 0.033 s per solve update.
So on the large problem the continuation is faster than the simple task group but it is still 30% slower than the TBB task implementation. 

Even though the result can be replicated with either task groups or more complicated use of one TBB, 30% performance lost between TBB and one TBB is quite troublesome. We have been using TBB for 10 years and refined our use of tasks to reach the current performance. Having our simulations taking 30% more time than today to process our customers scenes is quite a big issue.

I don't know what can be done to get more performance out of the task group. I doubt adding recycling would produce 30% gain.

JC

**Comment by pavelkumbrasev** [comment]
It is quite interesting results because in algorithms that we implemented we basically did the same thing old TBB did so I would not expect a big difference performance wise.
In a current approach 3 things a missing:

1. No task recycling.
2. No task bypassing.
3. Memory allocation with system allocator. (might be a problem because usually done under system lock)

1-2 can be done by `mutable` states and `return tg.defer(*this);`
3. You can try allocators that oneTBB provide like [scalable_allocator](https://spec.oneapi.io/versions/latest/elements/oneTBB/source/memory_allocation/scalable_allocator_cls.html) or [cache_aligned_allocator](https://spec.oneapi.io/versions/latest/elements/oneTBB/source/memory_allocation/scalable_allocator_cls.html)
 
 Could you please try this steps?

**Comment by JCLeprevost** [comment]
I will try the bypassing and recycling. But I found the recycling exemple you wrote very confusing as I am not familiar with the API tbh. Also I am not a big fan of using mutable and bypass the constantness. 

I can already knock down the allocator aspect because our core library has an implemented Allocator using TBB scalable_aligned_ ... allocators by default such as:
![image](https://user-images.githubusercontent.com/86110507/232027733-c4d2e584-42ae-41b0-af18-cdfd04a2b73c.png)
We then enforce New with this template: 
![image](https://user-images.githubusercontent.com/86110507/232027943-7df7cacf-a534-424b-adb2-06e705c4b954.png)
And any class we design is declared this way:
![image](https://user-images.githubusercontent.com/86110507/232028126-10f2baaa-c6d5-4dbe-a9a8-63f75182dc6b.png)
So nothing should go through the system allocator.

I will still experiment with this though. One thing I did not mention earlier is that on smaller problems the simple task group code is actually faster than the continuation one.

JC

**Comment by pavelkumbrasev** [comment]
> One thing I did not mention earlier is that on smaller problems the simple task group code is actually faster than the continuation one.

TBH, I faced with a similar problem and it caused by single `task_group` and shared reference counting and this is causing a lot of overhead.
I will try to think how to rewrite proposed solution.

**Comment by JCLeprevost** [comment]
I made sure the allocator question is out of the way: 
![image](https://user-images.githubusercontent.com/86110507/232032898-0bee82ae-6c0f-43fe-9a52-d1e32295619f.png)
![image](https://user-images.githubusercontent.com/86110507/232032938-8d06952e-4d96-4d66-ae02-034b0b39d35e.png)
Also it would fire badly on collision detection if we were not using the TBB scalable malloc. The system allocation produces a 1000% slowdown on benchmark, I can't miss that.

**Comment by JCLeprevost** [comment]
If I remember well there was 2 options to handle the time dependency between right left and next:
1- Recycle the current task 2 times. First time it spawns Right as its child and recycle itself as a child  to run Left. Then the second task recycle would recycle itself as Next.
2-  Recycle the current task only 1 time. Allocate Next as a continuation, spawn Right as a child of continuation and Recycle itself as a child of continuation to process Left. 
The performance difference was marginally better with the 2nd option that is the algorithm I posted here. But the code is way simpler and way more robust using recycle as child of continuation than using tricks to recycle the same task twice using a counter or an extra stage variable and having to bother about reference counting.

![image](https://user-images.githubusercontent.com/86110507/232039467-37e8c256-eaec-46fb-a560-be0cd9c3274a.png)
You got to admit that it raises questions when you compare the amount of code needed to replicate the performance of this using oneTBB API. Also the task TBB interface might have issues, There is nothing complicated in this code: spawn a root, wire a child to root recycle itself as a child of root.
We even managed to write a fully order deterministic collision detection performing broad phase, narrow phase and contacts generations with 3 levels of nested parallelism and 6 tree-tree traversal combinations for rigid, animated, and deformable objects including self collision with a single Spawn root and wait. I am going to cry if I can't replicate that feat of strength in one TBB :D

**Comment by pavelkumbrasev** [comment]
Do you mind moving this conversation to a email? (You can find mine in the profile)
I would like to help with this, I think it can speed up / simplify the process.

**Comment by JCLeprevost** [comment]
sure
